Prototype Compose on Bazel

adhiamboperes opened the pull request:
<!-- READ ME FIRST: Please fill in the explanation section below and check off every point from the Essential Checklist! -->
## Explanation
Add compose dependencies.

The main problems with this PR are:
1. The app builds but crashes on launch of HomeActivity with error: 
```
`java.lang.NoSuchMethodError: No virtual method setContent(Lkotlin/jvm/functions/Function0;)V in class Landroidx/compose/ui/platform/ComposeView; or its super classes (declaration of 'androidx.compose.ui.platform.ComposeView' appears in /data/app/org.oppia.android-AI4CVmqx0K4NgT9uR1VZnA==/base.apk)` 
that causes a crash because the method is responsible for displaying composable views.
```
2. The compose kt plugin seems not to work as expected.

Additional notes:
- I did not add gradle dependencies here.
- I have not added the code for the sticky header.

## Essential Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [ ] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [ ] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [ ] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [ ] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [ ] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [ ] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

## For UI-specific PRs only
<!-- Delete these section if this PR does not include UI-related changes. -->
If your PR includes UI-related changes, then:
- Add screenshots for portrait/landscape for both a tablet & phone of the before & after UI changes
- For the screenshots above, include both English and pseudo-localized (RTL) screenshots (see [RTL guide](https://github.com/oppia/oppia-android/wiki/RTL-Guidelines))
- Add a video showing the full UX flow with a screen reader enabled (see [accessibility guide](https://github.com/oppia/oppia-android/wiki/Accessibility-A11y-Guide))
- For PRs introducing new UI elements or color changes, both light and dark mode screenshots must be included
- Add a screenshot demonstrating that you ran affected Espresso tests locally & that they're passing

adhiamboperes committed:
Setup compose on kt 1.6

adhiamboperes committed:
Shift to ActivityResultContracts

BenHenning committed:
Merge branch 'upgrade-to-kotlin1.6' into prototype-compose-bazel-kt1.6

Conflicts:
	app/src/main/java/org/oppia/android/app/home/HomeFragmentPresenter.kt
	third_party/maven_install.json

BenHenning committed:
Merge branch 'migrate-to-newer-bazel-and-kotlin' into prototype-compose-bazel-kt1.6

Conflicts:
	third_party/maven_install.json

BenHenning commented:
NB: I updated this to be based on #4886 instead of the rules Kotlin branch since we need to make sure this specifically works with Bazel 6.x as that's the change that needs to get merged in to unblock the code coverage GSoC project.

BenHenning committed:
Fix app/BUILD.bazel & revert extra app changes.

BenHenning committed:
Attempt to update to newer Compose versions.

This puts our attempt in-line with the rules_kotlin example closest to
our version.

BenHenning commented:
Spent a bit of time investigating this, but will need to look more into it tomorrow.

So far, it seems ComposeView does have setContentView but it takes a Function0 not a Function2 (there might be another one, I only checked classes.dex in the built oppia_dev.aab file). I still need to dig into HomeFragmentPresenter.class to see if there's actually Compose code being generated by the plugin (since it's hard to tell if it's actually running), and I need to verify that rules_kotlin's version works for us, too: https://github.com/bazelbuild/rules_kotlin/blob/f07c7687c6a63058c2910c91f9f976a468085cb9/examples/jetpack_compose.

BenHenning commented:
Pretty sure the plugin isn't running; I commented out the plugins line in app/BUILD.bazel and it didn't change the build graph at all. Will dig into what's going on.

BenHenning commented:
Aha, the issue is that the android_library() we're using to wrap all third-party dependencies seems incompatible with kt_compiler_plugin(). Using the Maven target directly actually activates the plugin, though I'm still working to find the correct version list that will build with our Bazel/Kotlin setup. This is a reasonable workaround to figure out whether it'll work.

I've actually been thinking about removing our wrappers and using aliases instead. It should have better build performance and compatibility.

BenHenning commented:
Progress! Coursier (the utility that handles compiling the transitive closure of compatible Maven dependencies in rules_jvm_external) doesn't seem to handle beta/prerelease versions well, so we needed to find a non-prerelease version compatible with activity-compose version 1.4.0. This turned out to be 1.1.1. Unfortunately, the Compose compiler is _very_ picky about the Kotlin compiler version (https://www.jetbrains.com/help/kotlin-multiplatform-dev/compose-compatibility-and-versioning.html#kotlin-compatibility though some of these are wrong), so we needed to downgrade to 1.6.10 from 1.6.21.

From there, the Compose compiler seems to be running. Launching the app now results in a new failure:

```
05:18:07.862  E  FATAL EXCEPTION: main
                 Process: org.oppia.android, PID: 11582
                 java.lang.IllegalStateException: ViewTreeLifecycleOwner not found from androidx.drawerlayout.widget.DrawerLayout{e02b7f2 VFE...... ......I. 0,0-0,0 #7f0901cc app:id/home_activity_drawer_layout}
                 	at androidx.compose.ui.platform.WindowRecomposer_androidKt.createLifecycleAwareViewTreeRecomposer(WindowRecomposer.android.kt:244)
                 	at androidx.compose.ui.platform.WindowRecomposer_androidKt.access$createLifecycleAwareViewTreeRecomposer(WindowRecomposer.android.kt:1)
                 	at androidx.compose.ui.platform.WindowRecomposerFactory$Companion$LifecycleAware$1.createRecomposer(WindowRecomposer.android.kt:99)
                 	at androidx.compose.ui.platform.WindowRecomposerPolicy.createAndInstallWindowRecomposer$ui_release(WindowRecomposer.android.kt:155)
                 	at androidx.compose.ui.platform.WindowRecomposer_androidKt.getWindowRecomposer(WindowRecomposer.android.kt:230)
                 	at androidx.compose.ui.platform.AbstractComposeView.resolveParentCompositionContext(ComposeView.android.kt:244)
                 	at androidx.compose.ui.platform.AbstractComposeView.ensureCompositionCreated(ComposeView.android.kt:251)
                 	at androidx.compose.ui.platform.AbstractComposeView.onAttachedToWindow(ComposeView.android.kt:283)
                 	at android.view.View.dispatchAttachedToWindow(View.java:17445)
                 	at android.view.ViewGroup.dispatchAttachedToWindow(ViewGroup.java:3326)
                 	at android.view.ViewGroup.dispatchAttachedToWindow(ViewGroup.java:3333)
                 	at android.view.ViewGroup.dispatchAttachedToWindow(ViewGroup.java:3333)
                 	at android.view.ViewGroup.dispatchAttachedToWindow(ViewGroup.java:3333)
                 	at android.view.ViewGroup.dispatchAttachedToWindow(ViewGroup.java:3333)
                 	at android.view.ViewGroup.dispatchAttachedToWindow(ViewGroup.java:3333)
                 	at android.view.ViewGroup.dispatchAttachedToWindow(ViewGroup.java:3333)
                 	at android.view.ViewGroup.dispatchAttachedToWindow(ViewGroup.java:3333)
                 	at android.view.ViewGroup.dispatchAttachedToWindow(ViewGroup.java:3333)
                 	at android.view.ViewGroup.dispatchAttachedToWindow(ViewGroup.java:3333)
                 	at android.view.ViewRootImpl.performTraversals(ViewRootImpl.java:1668)
                 	at android.view.ViewRootImpl.doTraversal(ViewRootImpl.java:1392)
                 	at android.view.ViewRootImpl$TraversalRunnable.run(ViewRootImpl.java:6752)
                 	at android.view.Choreographer$CallbackRecord.run(Choreographer.java:911)
                 	at android.view.Choreographer.doCallbacks(Choreographer.java:723)
                 	at android.view.Choreographer.doFrame(Choreographer.java:658)
                 	at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:897)
                 	at android.os.Handler.handleCallback(Handler.java:790)
                 	at android.os.Handler.dispatchMessage(Handler.java:99)
                 	at android.os.Looper.loop(Looper.java:164)
                 	at android.app.ActivityThread.main(ActivityThread.java:6494)
                 	at java.lang.reflect.Method.invoke(Native Method)
                 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:438)
                 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:807)
```

adhiamboperes commented:
@BenHenning, 

> To fix the view tree issue, suggested to upgrade the version of androidx.appcompat:appcompat library to 1.3.0+ (currently its 1.0.2). And it works fine after making this change. 
> 
> The [release note](https://developer.android.com/jetpack/androidx/releases/appcompat#1.3.0-alpha01) also mentions that a support for ViewTreeLifecycleOwner was added from 1.3.0-alpha01 version. 

@theMr17 solved the issue on gradle this way.

BenHenning commented:
Ah yeah I just came across https://stackoverflow.com/q/66382502/3689782. Upgrading appcompat seems to work. I'm not keen at all these runtime dependency issues with Compose...hopefully the Robolectric tests can catch these problems otherwise it's going to introduce a lot of manual testing complexity for the team.

That being said, it works!

![image](https://github.com/oppia/oppia-android/assets/12983742/6c011bb2-ea70-42b3-8051-ea82752ca58e)

Another issue I've noticed is that sticky header (https://developer.android.com/reference/kotlin/androidx/compose/foundation/lazy/LazyListScope#stickyHeader(kotlin.Any,kotlin.Any,kotlin.Function1)) requires Compose 1.2.0 (https://developer.android.com/jetpack/androidx/releases/compose-foundation#1.2.0) which will require Kotlin 1.7.x. Going to see if I can also get the sticky header working since all this effort won't really matter if we need to upgrade Kotlin right away, anyway.

Edit: though maybe the documentation is wrong? ``stickyHeader`` _is_ being used in code, so maybe it will work as-is. Going to tinker a bit...

BenHenning commented:
Surprisingly, stickHeader seems to work just fine actuallyactually, without needing to update any versions. I expanded the prototype example to match more the example in the docs:

https://github.com/oppia/oppia-android/assets/12983742/d4bc0bb1-083f-41ff-a3e1-df36aa92ddfd

BenHenning commented:
Also separately: using alias() seems to work fine. I'll pull that forward to one of the Bazel branch PRs (not sure which one yet).

BenHenning committed:
Get Jetpack Compose working.

This fully demonstrates a working Bazel build, plus sticky headers
working without issue.

BenHenning commented:
I'll double check this branch once I pull some of the compatibility changes earlier into the chain, but everything seems to be working now.

BenHenning commented:
Also: wanted to mention that I ultimately found that android_library() was the culprit from https://github.com/bazelbuild/rules_kotlin/issues/319 suggesting kt_jvm_import() and considering that it may behave differently than android_library(). That was also the only obvious difference from the example setup in the rules_kotlin repo.

adhiamboperes commented:
>  [jetpack_compose_sticky_header_bazel_demo.mp4](https://github.com/oppia/oppia-android/assets/12983742/d4bc0bb1-083f-41ff-a3e1-df36aa92ddfd)

This looks pretty good! 

@theMr17, could you please pull these changes  and verify your dummy UI in https://github.com/theMr17/oppia-android/commit/673b75c030b56ebbbe232a347a5e23c391c444b2? Please confirm that you can build with Bazel.

BenHenning committed:
Merge branch 'migrate-to-newer-bazel-and-kotlin' into prototype-compose-bazel-kt1.6

Conflicts:
	third_party/maven_install.json

BenHenning commented:
Okay I pulled forward some of the extra changes needed here so now I think this PR is very close to the smallest number of changes needed to demonstrate working Jetpack Compose.

I also verified that the functionality seems to build and work correctly with a Proguard build (oppia_beta); surprisingly no new Proguard exemptions are needed which is nice to see.

De-assigning myself now since I don't think there's anything else for me to investigate at this point.

theMr17 commented:
> @theMr17, could you please pull these changes and verify your dummy UI in [theMr17@673b75c](https://github.com/theMr17/oppia-android/commit/673b75c030b56ebbbe232a347a5e23c391c444b2)? Please confirm that you can build with Bazel.

@adhiamboperes I pulled these changes and added the dummy UI. I can confirm everything works fine.


https://github.com/oppia/oppia-android/assets/84731134/3f20f3d7-9167-4fc6-877b-2d63684d4679

adhiamboperes commented on tools/kotlin/BUILD.bazel:
```diff
@@ -2,15 +2,15 @@
 kt_kotlinc_options(
     name = "oppia_kotlinc_options",
-    warn = "error",
+    warn = "report",
```
@BenHenning, is it intentional changing the WError to report here?

BenHenning replied:
As discussed in the CLaM meeting, this was just to simplify this prototype PR though I've made a note in the Kotlin upgrade PR to make this a bit more lenient for local development (but still strict for CI so that we can continue to keep build warnings at a minimum).

adhiamboperes commented:
Thanks @BenHenning, @theMr17! This is nice to see.
